### PR TITLE
[Transform] Add FeatMask and RowFeatNormalizer transforms

### PR DESCRIPTION
**Pull request by RecLusIve-F**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/transforms/module.py [hunk 1]
```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
     def __repr__(self):
         return self.__class__.__name__ + '()'
 
+class NormalizeFeatures(BaseTransform):
```
Comment (mufeili) [inline python/dgl/transforms/module.py:103]: Is `FeatNormalizer` a better name?

File: python/dgl/transforms/module.py [hunk 2]
```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
...
+        Default: `None`.
+
+    Example
+    -------
+
+    >>> import dgl
+    >>> import torch
+    >>> from dgl import NormalizeFeatures
+
+    Case1: Normalize features of a homogenous graph.
```
Comment (mufeili) [inline python/dgl/transforms/module.py:123]: It seems that people use `homogeneous` more often in the GNN field.

File: python/dgl/transforms/module.py [hunk 3]
```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
...
+    tensor([1.0000, 1.0000]) tensor([1.0000, 1.0000, 1.0000])
+    >>> print(new_g.edata['w'][('user', 'follows', 'user')].sum(1), new_g.edata['w'][('player', 'plays', 'game')].sum(1))
+    tensor([1.0000, 1.0000]) tensor([1.0000, 1.0000])
+    """
+    def __init__(self, node_feat_names=None, edge_feat_names=None):
+        self.node_feat_names = node_feat_names
+        self.edge_feat_names = edge_feat_names
+
+    def __call__(self, g):
+        def normalize(feat):
```
Comment (mufeili) [inline python/dgl/transforms/module.py:167]: You can add it as a class method like below:

```python
def normalize(self, feat):
    ...
```

File: python/dgl/transforms/module.py [hunk 4]
```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
...
+            feat = feat - feat.min()
+            feat.div_(feat.sum(dim=-1, keepdim=True).clamp_(min=1.))
```
Comment (mufeili) [inline python/dgl/transforms/module.py:169]: Is there a term for this kind of standardization? It does not seem to be a common standardization and it's better to rename the class to reflect the standardization approach.

File: python/dgl/transforms/module.py [hunk 5]
```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
...
+            return feat
+
+        if self.node_feat_names is None:
+            self.node_feat_names = g.ndata.keys()
```
Comment (mufeili) [inline python/dgl/transforms/module.py:173]: Does it work if nodes/edges of different types have different feature names?

File: python/dgl/transforms/module.py [hunk 6]
```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
...
+        for edge_feat_name in self.edge_feat_names:
+            if isinstance(g.edata[edge_feat_name], torch.Tensor):
+                g.edata[edge_feat_name] = normalize(g.edata[edge_feat_name])
+            else:
+                for etype in g.edata[edge_feat_name].keys():
+                    g.edges[etype].data[edge_feat_name] = normalize(g.edges[etype].data[edge_feat_name])
+
+        return g
+
+class NodeFeatureMasking(BaseTransform):
```
Comment (mufeili) [inline python/dgl/transforms/module.py:194]: Perhaps you can generalize this transform to edge features and rename the class accordingly. `FeatMask` or `FeatDrop` might be shorter.

File: python/dgl/transforms/module.py [hunk 7]
```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
...
+    r"""Randomly mask node features, as described in `An Empirical Study of Graph Contrastive Learning
+    <https://arxiv.org/abs/2109.01116>`__.
+
```
Comment (mufeili) [inline python/dgl/transforms/module.py:197]: It's better to mention that each time a column of the feature tensor is masked or not as a whole.

File: python/dgl/transforms/module.py [hunk 8]
```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
...
+    Parameters
+    ----------
+    p : float, optional
+        Probability of masking a node feature. Default: `0.5`.
```
Comment (mufeili) [inline python/dgl/transforms/module.py:201]: "a node feature" -> "a column of a feature tensor"?

Comment (mufeili) [inline python/dgl/transforms/module.py:201]: Add an argument to specify the name of the feature(s) to mask?

File: python/dgl/transforms/module.py [hunk 9]
```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
...
+
+    Example
+    -------
+
+    >>> import dgl
+    >>> import torch
+    >>> from dgl import NodeFeatureMasking
+
+    Case1 : Mask node features of a homogenous graph.
```
Comment (mufeili) [inline python/dgl/transforms/module.py:210]: Ditto for homogenous

File: python/dgl/transforms/module.py [hunk 10]
```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
...
+
+    >>> transform = NodeFeatureMasking()
+    >>> g = dgl.rand_graph(5, 20)
+    >>> g.ndata['h'] = torch.ones((g.num_nodes(), 10))
+
+    >>> new_g = transform(g)
```
Comment (mufeili) [inline python/dgl/transforms/module.py:216]: Perhaps you can still call it `g` as the transform does not return a new `DGLGraph` object.

File: tests/compute/test_transform.py [hunk 11]
```diff
@@ -2350,6 +2350,58 @@ def test_module_laplacian_pe(idtype):
     else:
         assert F.allclose(new_g.ndata['lappe'].abs(), tgt)
 
+@parametrize_dtype
+def test_module_normalize_features(idtype):
+    transform = dgl.NormalizeFeatures()
+
+    # Case1: Normalize features of a homogenous graph.
```
Comment (mufeili) [inline tests/compute/test_transform.py:2357]: Ditto for `homogenous`

File: tests/compute/test_transform.py [hunk 12]
```diff
@@ -2350,6 +2350,58 @@ def test_module_laplacian_pe(idtype):
...
+    new_g = transform(g)
+    assert F.allclose(new_g.ndata['h'].sum(1), F.tensor([1.0, 1.0, 1.0]))
+    assert F.allclose(new_g.edata['w'].sum(1), F.tensor([1.0, 1.0]))
+
+    # Case2: Normalize features of a heterogeneous graph.
+    g = dgl.heterograph({
+        ('user', 'follows', 'user'): (F.tensor([1, 2]), F.tensor([3, 4])),
+        ('player', 'plays', 'game'): (F.tensor([2, 2]), F.tensor([1, 1]))
+    })
+    g.ndata['h'] = {'game': F.randn(2, 5), 'player': F.randn(3, 5)}
```
Comment (mufeili) [inline tests/compute/test_transform.py:2370]: Add a feature `h2` for nodes of type `user`?

**Comment by mufeili** [comment]
The lint check failed:

```

python/dgl/transforms/module.py:110:0: C0301: Line too long (103/100) (line-too-long)

python/dgl/transforms/module.py:113:0: C0301: Line too long (103/100) (line-too-long)

python/dgl/transforms/module.py:153:0: C0301: Line too long (119/100) (line-too-long)

python/dgl/transforms/module.py:154:0: C0301: Line too long (113/100) (line-too-long)

python/dgl/transforms/module.py:159:0: C0301: Line too long (121/100) (line-too-long)

python/dgl/transforms/module.py:183:0: C0301: Line too long (104/100) (line-too-long)

python/dgl/transforms/module.py:190:0: C0301: Line too long (104/100) (line-too-long)

python/dgl/transforms/module.py:195:0: C0301: Line too long (102/100) (line-too-long)

python/dgl/transforms/module.py:255:0: C0301: Line too long (116/100) (line-too-long)
```

**Review by mufeili — CHANGES_REQUESTED** [review]
Thanks. Done a first pass.

File: python/dgl/transforms/module.py (hunk 2 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py:123]: Got it, I fixed it.

File: python/dgl/transforms/module.py (hunk 4 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py:169]: This normalisation is based on the [pyg implementation](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html#torch_geometric.transforms.NormalizeRotation). It normalized the sample features sum up to 1.

File: python/dgl/transforms/module.py (hunk 5 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py:173]: Yes, it does work. For heterogeneous graph, the `g.ndata` stores node data like this `{'feat_name': {'node_type': tensor}}`. So `g.ndata.keys()` will get all the feature names.

File: python/dgl/transforms/module.py (hunk 9 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py:210]: Fixed

File: python/dgl/transforms/module.py (hunk 10 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py:216]: Fixed

File: tests/compute/test_transform.py (hunk 12 above)
Comment (RecLusIve-F) [reply tests/compute/test_transform.py:2370]: Added

File: tests/compute/test_transform.py (hunk 11 above)
Comment (RecLusIve-F) [reply tests/compute/test_transform.py:2357]: Fixed

File: python/dgl/transforms/module.py (hunk 3 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py:167]: This function actually is a static method, is it proper to add it as a class method?

File: python/dgl/transforms/module.py (hunk 6 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py:194]: Generalized to `FeatDrop`, both node and edge features can transformed by this transform.

File: python/dgl/transforms/module.py (hunk 8 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py:201]: Fixed

File: python/dgl/transforms/module.py (hunk 7 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py:197]: Fixed

File: python/dgl/transforms/module.py (hunk 8 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py:201]: Fixed

File: python/dgl/transforms/module.py (hunk 1 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py]: Fixed

**Comment by RecLusIve-F** [comment]
I skipped testing on the mxnet and tensorflow backends due to `g.ndata` only storing data in `{'feat_name': {'node_type': tensor}}` format on the PyTorch backend and the use of `torch.Bernoulli` in the implementation.

File: python/dgl/transforms/module.py (hunk 4 above)
Comment (mufeili) [reply python/dgl/transforms/module.py]: I understand. However:

1) It's an optional step to do the subtraction before division.
2) There can be many other different approaches for feature normalization. So either we can rename this class or we need to make it compatible with other common normalization approaches.

**Comment by mufeili** [comment]
> I skipped testing on the mxnet and tensorflow backends due to `g.ndata` only storing data in `{'feat_name': {'node_type': tensor}}` format on the PyTorch backend and the use of `torch.Bernoulli` in the implementation.

I'm fine with that.

File: python/dgl/transforms/module.py (hunk 4 above)
Comment (mufeili) [reply python/dgl/transforms/module.py]: Also the PyG API you referred to seems to do the subtraction by the mean rather than minimal value.

Comment (RecLusIve-F) [reply python/dgl/transforms/module.py]: Sorry for I referred to the wrong PyG API. Here is the right [PyG API](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/transforms/normalize_features.html#NormalizeFeatures).To do subtraction by the minimal value is aiming to handle the negative values.

Comment (RecLusIve-F) [reply python/dgl/transforms/module.py]: I think making the class compatible with other common normalization methods is a good way to go, but renaming the class to better reflect the normalization would be an easy way to handle it for now. I can't find a good name that reflects the normalisation. Do you have any suggestions?

Comment (mufeili) [reply python/dgl/transforms/module.py]: How about calling it `RowFeatNormalizer` with an optional class argument `subtract_min`?

Comment (RecLusIve-F) [reply python/dgl/transforms/module.py]: I think it's a good name. When the `subtract_min` is an optional class argument, we can't make all rows of feature tensor sum up to 1 if there are negative values.

Comment (mufeili) [reply python/dgl/transforms/module.py]: That's fine. You can leave a note in the argument description of `subtract_min`.

File: python/dgl/transforms/module.py [hunk 13]
```diff
@@ -98,6 +100,255 @@ def __call__(self, g):
...
+class RowFeatNormalizer(BaseTransform):
+    r"""
+    Row-normalizes the features given in ``node_feat_names`` and ``edge_feat_names``.
+
+    The row normalization formular is:
+
+    .. math::
+      x = \frac{x}{\sum_i x_i}
+
+    where :math:`x` denotes the one row of the feature tensor.
```
Comment (mufeili) [inline python/dgl/transforms/module.py:112]: "the one" -> "a"

**Review by mufeili — APPROVED** [review]
LGTM other than a minor grammar issue.

File: python/dgl/transforms/module.py (hunk 13 above)
Comment (RecLusIve-F) [reply python/dgl/transforms/module.py]: fixed

**Comment by mufeili** [comment]
Thanks. I've merged the PR.

**Comment by chengpei-wu** [comment]
Based on my observations, it seems that FeatMask changes the node features of the original graph, which I think is somewhat unreasonable. NodeShuffle does not change the original node features. As shown in the example below.
```python
# feat mask
feat_mask = FeatMask(0.5, ['feat'])
g = dgl.rand_graph(10, 20)
g.ndata['feat'] = torch.arange(40).view(10, -1)
new_g = feat_mask(g)
print(new_g.ndata['feat'])
print(g.ndata['feat'])  # the same as new_g.ndata['feat']

# node shuffle
g = dgl.rand_graph(10, 20)
g.ndata['feat'] = torch.arange(40).view(10, -1)
aug = NodeShuffle()
new_g = aug(g)
print(new_g.ndata['feat'])
print(g.ndata['feat']) # remain unchanged
```
